### PR TITLE
async/await

### PR DESCRIPTION
**Pull request by proggeramlug**
Not perfect but a starting point. Also definitely not tested for all providers.

I need this for Twitter/X and so I thought I'll bite the bullet and make it async/await.

### CODE REVIEW

File: Sources/ImperialGitHub/GitHubRouter.swift [hunk 1]
```diff
@@ -16,7 +15,7 @@ public class GitHubRouter: FederatedServiceRouter {
         return headers
     }()
 
-    public required init(callback: String, completion: @escaping (Request, String) throws -> (EventLoopFuture<ResponseEncodable>)) throws {
+    public required init(callback: String, completion: @escaping (Vapor.Request, String) async throws -> Vapor.Response) async throws {
```
Comment (chibombo) [inline Sources/ImperialGitHub/GitHubRouter.swift:18]: I guess we don't need to add the `Vapor.` declaration, Should we keep it?

**Review by chibombo — CHANGES_REQUESTED** [review]

File: Sources/ImperialCore/Routing/FederatedServiceRouter.swift [hunk 2]
```diff
@@ -43,32 +43,32 @@ public protocol FederatedServiceRouter {
...
-    func callbackBody(with code: String) -> ResponseEncodable
+    func callbackBody(with code: String) -> any Content
     
     /// The route that the OAuth provider calls when the user has been authenticated.
     ///
     /// - Parameter request: The request from the OAuth provider.
     /// - Returns: A response that should redirect the user back to the app.
     /// - Throws: An errors that occur in the implementation code.
-    func callback(_ request: Request) throws -> EventLoopFuture<Response>
+    func callback(_ request: Request) async throws -> Response
```
[the remark this answers is not in the document]
Comment (0xTim) [reply Sources/ImperialCore/Routing/FederatedServiceRouter.swift:71]: `@Sendable` should not be added to function declarations - it's a half broken implementation that currently just disables sendable checking inside that function. We should definitely enable Sendable checking on the library though

**Comment by wibed** [comment]
any update?

**Comment by 0xTim** [comment]
Just an FYI I'm going to be doing an Imperial code sprint next week to go through all the issues and PRs

**Comment by bwdmr** [comment]
i did a small rewrite, in which i applied the pattern of the jwt-kit and layered oauth on top of it.
if youd like to have a look:
https://github.com/bwdmr/oauth-kit

**Comment by 0xTim** [comment]
Closing in favour of https://github.com/vapor-community/Imperial/pull/109

**Pull request closed without merging**